staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go

```diff
@@ -218,54 +224,161 @@ func (w *tunnelingResponseWriter) WriteHeader(statusCode int) {
 }
 
 // headerInterceptingConnection wraps the tunneling "net.Conn" to drain the
-// HTTP response from the upstream SPDY connection, since the client has
-// already received a "101 Switching Protocols" from the WebSockets
-// upgrade.
+// HTTP response status/headers from the upstream SPDY connection, then use
+// that to decide how to initialize the delegate connection for writes.
 type headerInterceptingConnection struct {
-	net.Conn
+	// initializableConn is delegated to for all net.Conn methods.
+	// initializableConn.Write() is not called until response headers have been read
+	// and initializableConn#InitializeWrite() has been called with the result.
+	initializableConn
+
+	lock         sync.Mutex
+	headerBuffer []byte
+	initialized  bool
+}
 
-	headerBuffer     *bytes.Buffer
-	completedHeaders bool
-	parsedStatus     string
-	parsedHeaders    http.Header
+// initializableConn is a connection that will be initialized before any calls to Write are made
+type initializableConn interface {
+	net.Conn
+	InitializeWrite(backendResponse *http.Response) error
 }
 
 // Write intercepts to initially swallow the HTTP response, then
 // delegate to the tunneling "net.Conn" once the response has been
 // seen and processed.
 func (h *headerInterceptingConnection) Write(b []byte) (int, error) {
-	if h.completedHeaders {
-		return h.Conn.Write(b)
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	if h.initialized {
+		return h.initializableConn.Write(b)
 	}
 
 	// Write into the headerBuffer, then attempt to parse the bytes
 	// as an http response.
-	n, err := h.headerBuffer.Write(b)
-	if err != nil {
-		return n, err
-	}
-	bufferedReader := bufio.NewReader(h.headerBuffer)
+	h.headerBuffer = append(h.headerBuffer, b...)
+	bufferedReader := bufio.NewReader(bytes.NewReader(h.headerBuffer))
 	resp, err := http.ReadResponse(bufferedReader, nil)
 	if errors.Is(err, io.ErrUnexpectedEOF) {
 		// don't yet have a complete set of headers
-		// TODO: reset headerBuffer to append to end and read from start
-		return n, nil
+		return len(b), nil
 	}
 	if err != nil {
 		klog.Errorf("invalid headers: %v", err)
-		return n, err
+		return len(b), err
 	}
+	resp.Body.Close()
```

Comment:
need to verify this is correct and doesn't drain the buffered reader (unit tests should exercise this)